tests/integration/integration_test.go

```diff
@@ -1669,3 +1671,69 @@ func TestGetSchemaUsesCorrectVersion(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, "3.6.0", packageSpec.Version)
 }
+
+// Regression test for https://github.com/pulumi/pulumi/issues/19905
+func TestComponentProviderErrorInResourceRegistration(t *testing.T) {
+	// We want to install the command provider
+	t.Setenv("PULUMI_DISABLE_AUTOMATIC_PLUGIN_ACQUISITION", "false")
+
+	// The regression caused a hang where a remote component construct would
+	// never return.
+	timeout := time.After(3 * time.Minute)
+	done := make(chan bool)
+	go func() {
+		pulumiHome := t.TempDir()
+		integration.ProgramTest(t, &integration.ProgramTestOptions{
+			NoParallel:      true, // We're modifying the env above
+			PulumiHomeDir:   pulumiHome,
+			Dir:             "component-error-resource",
+			RelativeWorkDir: "program",
+			PrepareProject: func(info *engine.Projinfo) error {
+				providerPath := filepath.Join(info.Root, "..", "provider")
+
+				// Install command provider
+				t.Setenv("PULUMI_DISABLE_AUTOMATIC_PLUGIN_ACQUISITION", "false")
+				cmd := exec.Command("pulumi", "plugin", "install", "resource", "command", "1.0.4")
+				cmd.Env = append(cmd.Environ(), "PULUMI_HOME="+pulumiHome)
+				out, err := cmd.CombinedOutput()
+				require.NoError(t, err, "%s failed with: %s", cmd.String(), string(out))
+
+				// Install the provider's dependencies
+				installNodejsProviderDependencies(t, providerPath)
+
+				// Add the provider to our project
+				cmd = exec.Command("pulumi", "package", "add", providerPath)
+				cmd.Dir = info.Root
+				cmd.Env = append(cmd.Environ(), "PULUMI_HOME="+pulumiHome)
+				out, err = cmd.CombinedOutput()
+				require.NoError(t, err, "%s failed with: %s", cmd.String(), string(out))
+
+				return nil
+			},
+			Quick:         true,
+			ExpectFailure: true,
+			ExtraRuntimeValidation: func(t *testing.T, stack integration.RuntimeValidationStackInfo) {
+				foundError := false
+				for _, event := range stack.Events {
+					if event.DiagnosticEvent != nil && event.DiagnosticEvent.Severity == "error" {
+						t.Logf("DiagnosticEvent.Message: %s", event.DiagnosticEvent.Message)
+						if strings.Contains(event.DiagnosticEvent.Message, "exiting with error") {
+							foundError = true
+						}
+					}
+				}
+				require.True(t, foundError, "expected to find an error in the stack events")
+			},
+		})
+
+		done <- true
+	}()
+
+	select {
+	case <-timeout:
+		t.Fatal("Test didn't finish in time")
+	case <-done:
+	}
```

Comment:
Kinda makes me wish program test had a built in timeout functionality :thinking: